docs/elastic.rst

```diff
@@ -0,0 +1,144 @@
+and go from the Horovod job without interrupting the training process.
+
+
+When to use elastic training
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- You are running an `autoscaling <https://en.wikipedia.org/wiki/Autoscaling>`__ job that may acquire more resources for training over tine.
+- Your job is running on preemptable or spot instances that may come and go with little warning.
+- Your nodes are unreliable and you want your job to continue training if some of the hosts fail.
+
+
+Requirements
+~~~~~~~~~~~~
+
+- Python >= 3.6
+- Horovod >= 0.20.0 with Gloo support (install Horovod using ``HOROVOD_WITH_GLOO=1`` to ensure it is installed)
+- A way to discover available hosts at runtime
+- TensorFlow >= 1.15 or PyTorch >= 1.0
+
+
+Modifying the training script with State Synchronization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The biggest difference when moving from normal distributed training to elastic training is the need to synchronize
+state among the workers when workers are added or removed from the job.
+
+All state that is to be synchronized among the workers should be encapsulated in an object deriving from
+``hvd.elastic.State``. This object encapsulates everything that needs to be in sync between workers, including
+model / optimizer state, current epoch, current batch index, etc.  It will be used for issuing callbacks to adjust
+learning rate, redistribute data, etc.
+
+In order for state to be recovered in the event of a failure, you must call ``state.commit()`` periodically.
+Typically, commits are made once per batch, but can be less frequent if saving state is costly and chance of node
+failure is low.
+
+On commit, if any workers are pending being added, then they will be added after the commit. A rollback will occur on
+active workers, and they will resync from the sync point.
+
+A decorator ``hvd.elastic.run`` is used to drive the elastic training loop.  Decorating a function in this way
+indicates that the function execution needs to be synchronized across workers.  When failure occurs or new worker is
+added, other workers will rollback to this point to resync state before resuming training.
+
+Because the sync function uses collective ops, and upon worker add the active workers will not reset from before this
+function, *no Horovod collective ops (broadcast, allreduce, etc.) can be called before this function*.
+
+If a worker fails during training, the other workers will catch a ``HorovodInternalError`` that is raised when
+attempting to perform the next collective operation.  This will trigger a rollback to the beginning of the
+synchronized function.
+
+Rollback process is as follows:
+
+1. Catch exception within the hvd.elastic.run decorator.
+2. Restore last committed state (to clear any half-updated parameters).
+3. Reinitialize Horovod context performing a new round of rendezvous.
+4. Synchronize state among the workers by broadcasting from the new worker-0.
```

Comment:
One of the assumptions we make is that the driver may exist on different hardware from the workers.  As such, it could have a relatively slow NIC, no GPU, etc.  So we want to avoid sending data to the driver if possible.